[WF-569] Component documentation improvements

ixTec opened the pull request:
# Various documentation improvements

## [WF-569](https://datacamp.atlassian.net/browse/WF-569)

## Changes

### 🔧 Minor Changes

- Updated _best practices_ and _usage guidelines_ the following components:
    - Asset
    - Avatar
    - Badge
    - Brand
    - Button
    - Code
    - Code Block
    - Empty State
    - Heading
    - Icon
    - Loader
    - Paragraph
    - Progress
    - Tooltip

### ❓ Misc Changes

- Fixed _best-practices_ section margin when directly preceded by a `h3`
- Fixed typo in **Tooltip** `complex-content` example 
- Fixed margin issue with having a `Notification` at the top of a docs page
- Updated `new` flags for older components in docs navigation
- Updated roadmap

blurbyte commented on doc-site/pages/components/loader.mdx:
```diff
@@ -12,22 +13,46 @@ import Inverted from '../../examples/loader/inverted';
+  <BestPractices.Recommended>
+
+    - Use when an action will take more than a few seconds to complete
+    - Provide a simple but descriptive `aria-label` as context for screen readers&mdash;such as 'Loading user profile'
```
Shouldn't we make aria label mandatory? Also usage of `aria-live` / `aria-busy` in one of the examples should be great.

ixTec replied:
Agreed, `aria-label` should probably be mandatory. I have raised https://datacamp.atlassian.net/browse/WF-669 for this

blurbyte commented on doc-site/pages/components/paragraph.mdx:
```diff
@@ -11,11 +12,32 @@ import Variants from '../../examples/paragraph/variants';
 import Inverted from '../../examples/paragraph/inverted';
 
-<Layout title="Paragraph" description="🚧 This page is under construction">
+<Layout title="Paragraph" description="Used for styling blocks of text using DataCamp's styling.">
```
Maybe 'Used for styling blocks of text'. I would also mention here that's the default / go-to component for text.

blurbyte commented on doc-site/pages/components/paragraph.mdx:
```diff
@@ -11,11 +12,32 @@ import Variants from '../../examples/paragraph/variants';
 ## Usage guidance
 
-To highlight the content of importance, use regular `strong` or `em` elements.
+he `Paragraph` component offers three (3) different `size` values to choose from: `small`, `medium` and `large`. In most cases, the default `size` should be used.
```
Missing 'T', not a fan of (3) 😄

blurbyte commented on doc-site/pages/components/progress.mdx:
```diff
@@ -15,7 +16,42 @@ import CustomLabel from '../../examples/progress/custom-label';
+
+The `Progress` component can be used in a variety of uses; from showing the completion percentage of a task, to individual steps in a process. The `Progress` should be used when a process is a direct result of a user action&mdash;e.g. courses being completed or form progress.
+
+When using the default `mode`, any values for `min` and `max` can be provided and a percentage is calculated internally. This includes values that would result in a percentage over `100%`. This percentage is also always **rounded down**.
```
Maybe add 'Default continuous mode' heading above.

blurbyte commented on doc-site/pages/components/tooltip.mdx:
```diff
@@ -12,7 +13,54 @@ import WithComplexContent from '../../examples/tooltip/with-complex-content';
+  <BestPractices.Recommended>
+
+    - use with icon-only [Buttons](/components/button)
+    - use `Help`/`HelpCircle` [Icon](/components/icon) where trigger is not related
```
I think it should be icon-only button with `Help` / `HelpCircle` icon. But yeah this whole pattern is a bit sus.

ixTec replied:
Updated but yeah, it's a bit of a unclear use case as we don't really have a recommended standard yet (but this would be _my_ recommendation at least 😄 )

blurbyte reviewed (changes requested):
Looks great🥇!

A few things to consider to make it a bit more readable / consistent:
- in general I would avoid using bold / strong and use emphasis instead; but there are a few places when bold makes sense, like in 'critical' in Heading best practices
- not a big fan of number in parenthesis, e.g. something three (3), either omit it or use a number
- in BestPractices list items start lower case (most of them) or upper case (Avatar)?
- single quotes or double quotes: 'Course progress' or "Course progress"

ixTec commented on doc-site/pages/components/badge.mdx:
```diff
@@ -10,7 +11,38 @@ import playgroundConfig from '../../examples/badge/playground-config';
+
+The colors of a `Badge` should never be overwritten, as they are chosen for suitable accessibility and brand requirements.
+
+They should **never** be interactive and instead a [Button](/components/button) or the future `Tag` component used if interaction is desired.
```
Hopeful optimism from me here 😄

ixTec commented on doc-site/pages/components/icon.mdx:
```diff
@@ -9,63 +9,54 @@ import Sizes from '../../examples/icon/sizes';
-</BestPractices>
+An `Icon` is an image only, therefore by default it doesn't have any label or title and is hidden from assistive technologies, such as screen readers, by default. If a description for assistive technologies is necessary, use the `aria-label` attribute.
 
-### Icon is not a Button
```
I have it in the same place as before (as a heading in Best Practices), below the general best practices

ixTec commented:
- Updated emphasis across changes. I've kept usage for strong words such as "never", "not" and "critical", just to force the point more.
- Removed numbers in parenthesis 
- Double quotes should be used because we're following the US standards for grammar etc, instead of UK English (where it should be single quotes in most places)

blurbyte commented on doc-site/pages/components/code-block.mdx:
```diff
@@ -11,7 +12,32 @@ import BasicUsage from '../../examples/code-block/basic-usage';
+
+There are three `sizes` for `CodeBlock`, but in most cases the default value should be used. `CodeBlock` also has an `inverted` option for use with darker backgrounds.
+
+It should never be placed inline and instead the [Code](/components/code) component should be used.
```
Most likely 'never' should be bold here.

blurbyte commented on doc-site/pages/components/loader.mdx:
```diff
@@ -9,25 +10,53 @@ import Imports from '../../components/imports';
   description={
     <>
-      If you're looking for a loading button variant, see our{' '}
+      If you're looking for a loading Button variant, see our{' '}
```
Don't think button here should be capitalized here, since it's just generic use case of loader inside button.

ixTec replied:
Good point. I think it can be viewed in either way, but I believe the stronger wording is towards a generic button so I'll update this

blurbyte commented on doc-site/pages/components/loader.mdx:
```diff
@@ -9,25 +10,53 @@ import Imports from '../../components/imports';
+  <BestPractices.Recommended>
+
+    - use when an action will take more than a few seconds to complete
+    - provide a simple but descriptive `aria-label` as context for screen readers&mdash;such as 'Loading user profile'
```
Should be "Loading user profile"?

blurbyte commented on doc-site/pages/components/paragraph.mdx:
```diff
@@ -11,11 +12,32 @@ import Variants from '../../examples/paragraph/variants';
-To highlight the content of importance, use regular `strong` or `em` elements.
+The `Paragraph` component offers three different `size` values to choose from: `small`, `medium` and `large`. In most cases, the default `size` should be used.
+
+To highlight the content of importance, use regular `strong` or `em` elements. `Paragraph` also offers a more subtle `secondary` `variant` for use with less important or supportive content. On top of this, others components such as [Code](/components/code) and [Link](/components/link) can be used within the content.
```
Maybe `variant="secondary"`, will be a bit nicer to look at.

blurbyte commented on doc-site/examples/loader/aria-live.tsx:
```diff
@@ -0,0 +1,67 @@
+  </Paragraph>
+);
+
+function Example() {
```
It could be written in a bit more straightforward / cleaner way using only 1 state var instead of 3 (and trying to sync between them):

- for state use something along the lines `const [status, setStatus] = useState<'initial' | 'loading' | 'final'>('initial')` (or something similar)
- for content you can create `renderContent()` function with switch inside rendering appropriate content based on status
- for button just derive functionality / label from status

blurbyte reviewed (changes requested):
Nice updates 👏 . There are still a few things to improve:

- think we wanted to capitalize component name in props docs too (as in `Button`)
- still missing 'Icon is not a Button' heading in `Icon`
- for some components items on the `BestPractices` list start with uppercase and for others lowercase - would be nice to make it consistent
- `aria-live` code example for `Loader` could be improved, and to make content look a bit nicer maybe put it on beige `beigeLight` with rounded corners (as in `Resizable` examples)

ixTec commented:
Updated based on changes.

- I've replaced "Icon is not a Button" header with "Icon within a Button", as the text within this section clarifies that an Icon is not a Button in addition to the rest. I think this is a bit more suitable for the approach we're taking for the docs
- Updated casing of component name in prop docs for all components, but won't be tackling updating the other component's documentation beyond that in this PR (will be future work for either of us, as discussed)

blurbyte commented on doc-site/pages/components/paragraph.mdx:
```diff
@@ -11,11 +12,32 @@ import Variants from '../../examples/paragraph/variants';
-To highlight the content of importance, use regular `strong` or `em` elements.
+The `Paragraph` component offers three different `size` values to choose from: `small`, `medium` and `large`. In most cases, the default `size` should be used.
+
+To highlight the content of importance, use regular `strong` or `em` elements. `Paragraph` also offers a more subtle `variant="secondary`" for use with less important or supportive content. On top of this, others components such as [Code](/components/code) and [Link](/components/link) can be used within the content.
```
Quotation mark should be inside ticks for `variant="secondary"`.

blurbyte commented on doc-site/examples/loader/aria-live.tsx:
```diff
@@ -0,0 +1,89 @@
+    }
+  }
+
+  return (
```
I think it would look a bit cleaner (and no extra styles for button will be required):
<img width="857" alt="Screenshot 2022-10-20 at 11 13 33" src="https://user-images.githubusercontent.com/20565536/196908138-30d6764e-be63-435b-8676-cfe99a661bbb.png">

blurbyte replied:
I'm not sure about this rule tbh. It makes sense for Dialog (maybe for Card), but otherwise it's close to being anti-pattern, since users tend to scan long pages from left to right, therefore you want to put most important CTAs to the left.

blurbyte reviewed (changes requested):
Looks great, love how consistent it is ❤️ , and `aria-live` example is superb.
I have a few minor remarks, otherwise we should be ready to merge today.

<iframe src="https://giphy.com/embed/8Iv5lqKwKsZ2g" width="480" height="201" frameBorder="0" class="giphy-embed" allowFullScreen></iframe><p><a href="https://giphy.com/gifs/dicaprio-draw-serial-8Iv5lqKwKsZ2g">via GIPHY</a></p>

blurbyte commented on doc-site/examples/loader/aria-live.tsx:
```diff
@@ -0,0 +1,89 @@
+
+type ContentStatus = 'initial' | 'loading' | 'final';
+
+const INITIAL_CONTENT = (
```
Don't think it's worth extracting these to separate consts, I would put them directly in switch.

ixTec replied:
I think our examples should best practices in general where possible, so I've put these as consts so that they're not redefined each rerender when it's static content

blurbyte replied:
Makes sense.

blurbyte reviewed (approved):
Let's merge 🚀.
